CMakeLists.txt

```diff
@@ -53,25 +56,101 @@ One or both of these combinations are OFF in your installation but at least one
     )
 ENDIF()
 
+### GTEST/GMOCK  ######################################################
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}")
+
+set(GTEST_ROOT "/usr")
+find_package(GTest REQUIRED 1.8.0)
+find_package(GMock REQUIRED)
+#set(GTEST_INCLUDE_DIRS "/usr/include")
+#set(GTEST_BOTH_LIBRARIES "/usr/lib/libgtest_main.so;/usr/lib/libgtest.so")
+#set(GMOCK_BOTH_LIBRARIES "/usr/lib/libgmock.a;/usr/lib/libgmock_main.a")
+
+### COMPILE AND LINKER FLAGS #########################################
+
+SET(DEAL_II_CXX_FLAGS_DEBUG "${DEAL_II_CXX_FLAGS_DEBUG} -g -O0 --coverage")
+SET(DEAL_II_LINKER_FLAGS_DEBUG "${DEAL_II_LINKER_FLAGS_DEBUG} --coverage")
+
+######################################################################
+
 DEAL_II_INITIALIZE_CACHED_VARIABLES()
-SET(CLEAN_UP_FILES *.log *.gmv *.gnuplot *.gpl *.eps *.pov *.vtk *.ucd *.d2 *.vtu *.pvtu)
-PROJECT(xtrans)
+SET(CLEAN_UP_FILES *.log *.gmv *.gnuplot *.gpl *.eps *.pov *.vtk *.ucd
+  *.d2 *.vtu *.pvtu)
 ENABLE_TESTING()
 
-ADD_SUBDIRECTORY(src/aqdata)
-ADD_SUBDIRECTORY(src/common)
-ADD_SUBDIRECTORY(src/mesh)
+### EXECUTABLES AND SOURCES ##########################################
+
+# Recursively find source files
+
+file(GLOB_RECURSE sources "src/*.cpp" "src/*.cc")
+set(testing_sources ${sources})
+list(FILTER sources EXCLUDE REGEX ".(/tests/).")
+list(FILTER sources EXCLUDE REGEX ".(/test_helpers/).")
+
+# Include directories
+include_directories(${GTEST_INCLUDE_DIRS}
+  ${CMAKE_SOURCE_DIR}/inc)
+
+# Add xtrans executables
+ADD_EXECUTABLE(xtrans ${sources})
+ADD_EXECUTABLE(xtrans_test ${testing_sources})
+
+# Add testing definition and library to xtrans_test
+target_compile_definitions(xtrans_test PUBLIC -DTEST)
+
+target_link_libraries(xtrans_test ${GTEST_BOTH_LIBRARIES}
+  ${GMOCK_BOTH_LIBRARIES})
+
+ADD_SUBDIRECTORY(${CMAKE_SOURCE_DIR}/src/aqdata)
+ADD_SUBDIRECTORY(${CMAKE_SOURCE_DIR}/src/common)
+ADD_SUBDIRECTORY(${CMAKE_SOURCE_DIR}/src/mesh)
 SET(TARGET "xtrans")
-SET(TARGET_SRC src/main.cc)
+SET(TARGET_SRC ${CMAKE_SOURCE_DIR}/src/main.cc)
 
 # Add unit testing subdirectories
-ADD_SUBDIRECTORY(tests/aqdata)
-ADD_SUBDIRECTORY(tests/common)
-ADD_SUBDIRECTORY(tests/mesh)
-
-DEAL_II_INVOKE_AUTOPILOT()
-TARGET_LINK_LIBRARIES(xtrans
-_aqdata
-_common
-_mesh
-)
+ADD_SUBDIRECTORY(${CMAKE_SOURCE_DIR}/tests/aqdata)
+ADD_SUBDIRECTORY(${CMAKE_SOURCE_DIR}/tests/common)
+ADD_SUBDIRECTORY(${CMAKE_SOURCE_DIR}/tests/mesh)
+
```

Comment:
Nice use of CMAKE macro!

If the following lines are not useful, would you mind deleting them?